File: data/src/main/kotlin/dev/yashgarg/kimai/api/KimaiRepositoryImpl.kt
/*
 * Copyright © 2023 Yash Garg.
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */
package dev.yashgarg.kimai.api

import com.deliveryhero.whetstone.app.ApplicationScope
import com.squareup.anvil.annotations.ContributesBinding
import dev.yashgarg.kimai.models.Activity
import dev.yashgarg.kimai.models.Customer
import dev.yashgarg.kimai.models.Ping
import dev.yashgarg.kimai.models.TimesheetActivity
import dev.yashgarg.kimai.models.TimesheetBody
import dev.yashgarg.kimai.util.ApiResult
import dev.yashgarg.kimai.util.handleApi
import javax.inject.Inject

@ContributesBinding(ApplicationScope::class)
class KimaiRepositoryImpl @Inject constructor(private val api: KimaiApi) : KimaiRepository {
  override suspend fun ping(): ApiResult<Ping> = handleApi { api.ping() }

  override suspend fun getActivities(
    projectId: Int?,
    projects: List<Int>,
    visible: Int,
    globals: Boolean?,
    orderBy: String?,
    order: String,
    searchTerm: String?
  ): ApiResult<List<Activity>> = handleApi {
    api.getActivities(
      projectId = projectId,
      projects = projects,
      visible = visible,
      globals = globals,
      orderBy = orderBy,
      order = order,
      searchTerm = searchTerm
    )
  }

  override suspend fun getProjects(visible: Int): ApiResult<List<Activity>> = handleApi {
    api.getProjects(visible = visible)
  }

  override suspend fun getTimeSheets(
    orderBy: String?,
    order: String?
  ): ApiResult<List<TimesheetActivity>> = handleApi {
    api.getTimesheets(orderBy = orderBy, order = order)
  }

  override suspend fun getCustomers(visible: Int): ApiResult<List<Customer>> = handleApi {
    api.getCustomers(visible = visible)
  }

  override suspend fun createTimeSheet(
    begin: String,
    end: String?,
    project: Int,
    activity: Int
  ): ApiResult<TimesheetActivity> = handleApi {
    api.createTimesheet(
      TimesheetBody(begin = begin, end = end, project = project, activity = activity)
    )
  }
}

File: lib/data/api/kimai_repository_impl.dart
import 'package:dio/dio.dart';
import 'package:fpdart/fpdart.dart';
import 'package:injectable/injectable.dart';
import 'package:kimai/data/api/api_paths.dart';
import 'package:kimai/data/api/kimai_repository.dart';
import 'package:kimai/data/models/auth_user.dart';
import 'package:kimai/data/models/models.dart';

@LazySingleton(as: KimaiRepository)
class KimaiRepositoryImpl implements KimaiRepository {
  final Dio _dio;

  KimaiRepositoryImpl({
    required Dio dio,
  }) : _dio = dio;

  @override
  Future<Either<Ping, Exception>> ping(AuthUser user) async {
    try {
      final url = '${user.baseUrl}${ApiPaths.PING}';
      final headers = {
        'X-AUTH-USER': user.username,
        'X-AUTH-TOKEN': user.token,
      };

      final response = await _dio.get(
        url,
        options: Options(headers: headers),
      );

      return left(Ping.fromJson(response.data));
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<TimesheetActivity, Exception>> createTimeSheet({
    required String begin,
    String? end,
    required int project,
    required int activity,
  }) async {
    try {
      final response = await _dio.post(
        ApiPaths.GET_TIMESHEETS,
        data: {
          'begin': begin,
          if (end != null) 'end': end,
          'project': project,
          'activity': activity,
        },
      );

      return left(TimesheetActivity.fromJson(response.data));
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<List<Activity>, Exception>> getActivities({
    int? projectId,
    List<int>? projects,
    int visible = 1,
    bool? globals,
    String? orderBy,
    String order = 'ASC',
    String? searchTerm,
  }) async {
    try {
      final response = await _dio.get(
        ApiPaths.GET_ACTIVITIES,
        queryParameters: {
          if (projectId != null) 'project': projectId,
          if (projects != null) 'projects': projects,
          if (globals != null) 'globals': globals,
          if (orderBy != null) 'orderBy': orderBy,
          if (searchTerm != null) 'searchTerm': searchTerm,
          'order': order,
          'visible': visible,
        },
      );

      return left(
        (response.data as List).map((e) => Activity.fromJson(e)).toList(),
      );
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<List<Customer>, Exception>> getCustomers({
    int visible = 1,
  }) async {
    try {
      final response = await _dio.get(
        ApiPaths.GET_CUSTOMERS,
        queryParameters: {
          'visible': visible,
        },
      );

      return left(
        (response.data as List).map((e) => Customer.fromJson(e)).toList(),
      );
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<List<Activity>, Exception>> getProjects({
    int visible = 1,
  }) async {
    try {
      final response = await _dio.get(
        ApiPaths.GET_PROJECTS,
        queryParameters: {
          'visible': visible,
        },
      );

      return left(
        (response.data as List).map((e) => Activity.fromJson(e)).toList(),
      );
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<List<TimesheetActivity>, Exception>> getTimeSheets({
    String? orderBy,
    String? order,
  }) async {
    try {
      final response = await _dio.get(
        ApiPaths.GET_TIMESHEETS,
        queryParameters: {
          'orderBy': orderBy,
          'order': order,
        },
      );

      return left(
        (response.data as List)
            .map((e) => TimesheetActivity.fromJson(e))
            .toList(),
      );
    } on Exception catch (e) {
      return right(e);
    }
  }

  @override
  Future<Either<UserEntity, Exception>> getMyUser() async {
    try {
      final response = await _dio.get(ApiPaths.GET_ME);

      return left(UserEntity.fromJson(response.data));
    } on Exception catch (e) {
      return right(e);
    }
  }
}
